s4/clarity/_internal/factory.py

```diff
@@ -43,32 +43,73 @@ class ElementFactory(object):
     def _strip_params(string):
         return ElementFactory._params_re.sub('', string)
 
-    def __init__(self, lims, element_class, batch_flags=None, request_path=None, name_attribute="name"):
+    def __init__(self, lims, element_class):
         """
+        Creates a new factory to provide interface between Clarity LIMS and the client software.
+        The provided `element_class` will be used to convert the XML records into Python objects
+        and back. There are several class attributes that can be set to configure this interface.
+
+        NAME_ATTRIBUTE: This is used if the name of the class differs from the name in the Clarity API.
+
+        REQUEST_PATH: If the class has the REQUEST_PATH attribute then it will
+        be used to override the default path. It is used in relation to the root URI of the Clarity API.
+
+        BATCH_FLAGS: Batch Flags use the BitField enumeration to create a bit field that is used
+        to determine what batch services are provided for this data type. It defaults to None.
+
         :type lims: LIMS
         :type element_class: classobj
-        :type batch_flags: BatchFlags or None
-        :type request_path: str
-        :param request_path: for example, '/configuration/workflows'.
-                             when not specified, uses '/<plural of element name>'.
-        :type name_attribute: str
-        :param name_attribute: if not "name", provide this to adjust behaviour of 'get_by_name'.
         """
 
         self.lims = lims
         self.element_class = element_class
-        self.name_attribute = name_attribute
-        self.batch_flags = batch_flags or BatchFlags.NONE
+        self.name_attribute = self._get_name_attribute()
+        self.batch_flags = self._get_batch_flag()
         self._plural_name = self.element_class.__name__.lower() + "s"
-
-        if request_path is None:
-            request_path = "/" + self._plural_name
-        self.uri = lims.root_uri + request_path
+        self.uri = self._get_endpoint_uri()
 
         self._cache = dict()
 
         lims.factories[element_class] = self
 
+    def _get_name_attribute(self):
+        if hasattr(self.element_class, 'NAME_ATTRIBUTE'):
+            return self.element_class.NAME_ATTRIBUTE
+        return "name"
```

Comment:
This whole thing could just be `return getattr(self.element_class, 'NAME_ATTRIBUTE', "name")`. (I also don't know why it's a function, since why would anyone subclass and override it, but I guess if you really want it to be that's fine)

Reply:
They were separated out to remove the logic of determining where the name attribute was to be found from the logic of the constructor.
I really like the change you have of passing the default to getattr, that is small enough to put inline into the init.